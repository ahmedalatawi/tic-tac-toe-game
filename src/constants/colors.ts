export const Colors = {
  gray: "#a9a9a9",
  grayLight: "#DDDDDD",
  white: "#FFFFFF",
  black: "#000000",
};
