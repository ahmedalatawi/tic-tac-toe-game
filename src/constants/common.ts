export const BOARD_SIZES = {
  _3x3: 3,
  _4x4: 4,
  _5x5: 5,
};

export const PLAYER_X = 1;
export const PLAYER_O = 2;
export const TIE = 0;

export const PLAYER_TYPES = {
  user: "user",
  computer: "computer",
};

export const GAME_STATES = {
  notStarted: "not_started",
  inProgress: "in_progress",
  ended: "ended",
};
