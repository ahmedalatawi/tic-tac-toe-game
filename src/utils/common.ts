import {
  TIE,
  PLAYER_O,
  PLAYER_X,
  PLAYER_TYPES,
  BOARD_SIZES,
} from "../constants/common";
import { GameBoard, GameSettings } from "../types/types";

export const createBoard = (size: number) =>
  new Array(size).fill(Array(size).fill(null));

// Get a random number between min and max
export const getRandomNumber = (min: number, max: number) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1)) + min;
};

export const switchPlayer = (player: number) => {
  return player === PLAYER_X ? PLAYER_O : PLAYER_X;
};

/**
 * This function is used to return a random move for the computer
 * @param board
 * @param boardSize
 * @returns
 */
export const getRandomMove = (board: GameBoard, boardSize?: number | null) => {
  const max = (boardSize ?? BOARD_SIZES._3x3) - 1;
  let row = getRandomNumber(0, max);
  let col = getRandomNumber(0, max);

  while (board[row][col]) {
    row = getRandomNumber(0, max);
    col = getRandomNumber(0, max);
  }

  return [row, col];
};

/**
 * This function returns the game's final status based on player X and O, or tie
 * @param winner
 * @returns
 */
export const getWinnerText = (winner: number) => {
  let winnerText = "";
  switch (winner) {
    case PLAYER_X:
      winnerText = "The winner is player X";
      break;
    case PLAYER_O:
      winnerText = "The winner is player O";
      break;
    case TIE:
    default:
      winnerText = "It's a tie";
  }

  return winnerText;
};

/**
 * This function calculates all winning possibilities, e.g. column, row, or diagonal in order to determine the winner
 *
 * TODO: This function can def be improved by calculating all possibilities dynamically - no hard coding
 * @param board
 * @param boardSize
 * @returns
 */
export const getWinner = (board: GameBoard, boardSize?: number | null) => {
  const dimensions = boardSize ?? BOARD_SIZES._3x3;

  const rowMarked = (row: number, board: GameBoard) => {
    if (dimensions === BOARD_SIZES._3x3) {
      return (
        board[row][0] !== null &&
        board[row][0] === board[row][1] &&
        board[row][1] === board[row][2]
      );
    }
    if (dimensions === BOARD_SIZES._4x4) {
      return (
        board[row][0] !== null &&
        board[row][0] === board[row][1] &&
        board[row][1] === board[row][2] &&
        board[row][2] === board[row][3]
      );
    }
    if (dimensions === BOARD_SIZES._5x5) {
      return (
        board[row][0] !== null &&
        board[row][0] === board[row][1] &&
        board[row][1] === board[row][2] &&
        board[row][2] === board[row][3] &&
        board[row][3] === board[row][4]
      );
    }
  };

  const colMarked = (column: number, board: GameBoard) => {
    if (dimensions === BOARD_SIZES._3x3) {
      return (
        board[0][column] !== null &&
        board[0][column] === board[1][column] &&
        board[1][column] === board[2][column]
      );
    }
    if (dimensions === BOARD_SIZES._4x4) {
      return (
        board[0][column] !== null &&
        board[0][column] === board[1][column] &&
        board[1][column] === board[2][column] &&
        board[2][column] === board[3][column]
      );
    }
    if (dimensions === BOARD_SIZES._5x5) {
      return (
        board[0][column] !== null &&
        board[0][column] === board[1][column] &&
        board[1][column] === board[2][column] &&
        board[2][column] === board[3][column] &&
        board[3][column] === board[4][column]
      );
    }
  };

  const diagonalMarked = (diagonal: number, board: GameBoard) => {
    if (dimensions === BOARD_SIZES._3x3) {
      if (diagonal === 1) {
        return (
          board[0][0] !== null &&
          board[0][0] === board[1][1] &&
          board[1][1] === board[2][2]
        );
      } else {
        return (
          board[2][0] !== null &&
          board[2][0] === board[1][1] &&
          board[1][1] === board[0][2]
        );
      }
    }
    if (dimensions === BOARD_SIZES._4x4) {
      if (diagonal === 1) {
        return (
          board[0][0] !== null &&
          board[0][0] === board[1][1] &&
          board[1][1] === board[2][2] &&
          board[2][2] === board[3][3]
        );
      } else {
        return (
          board[2][0] !== null &&
          board[3][0] === board[2][1] &&
          board[2][1] === board[1][2] &&
          board[1][2] === board[0][3]
        );
      }
    }
    if (dimensions === BOARD_SIZES._5x5) {
      if (diagonal === 1) {
        return (
          board[0][0] !== null &&
          board[0][0] === board[1][1] &&
          board[1][1] === board[2][2] &&
          board[2][2] === board[3][3] &&
          board[3][3] === board[4][4]
        );
      } else {
        return (
          board[2][0] !== null &&
          board[4][0] === board[3][1] &&
          board[3][1] === board[2][2] &&
          board[2][2] === board[1][3] &&
          board[1][3] === board[0][4]
        );
      }
    }
  };

  if (dimensions === BOARD_SIZES._5x5 && rowMarked(4, board)) {
    return board[4][0];
  }
  if (dimensions === BOARD_SIZES._4x4 && rowMarked(3, board)) {
    return board[3][0];
  }

  if (dimensions === BOARD_SIZES._5x5 && colMarked(4, board)) {
    return board[0][4];
  }
  if (dimensions === BOARD_SIZES._4x4 && colMarked(3, board)) {
    return board[0][3];
  }

  if (dimensions === BOARD_SIZES._5x5 && diagonalMarked(2, board)) {
    return board[4][0];
  }
  if (dimensions === BOARD_SIZES._4x4 && diagonalMarked(2, board)) {
    return board[3][0];
  }

  if (rowMarked(0, board)) {
    return board[0][0];
  }
  if (rowMarked(1, board)) {
    return board[1][0];
  }
  if (rowMarked(2, board)) {
    return board[2][0];
  }

  if (colMarked(0, board)) {
    return board[0][0];
  }
  if (colMarked(1, board)) {
    return board[0][1];
  }
  if (colMarked(2, board)) {
    return board[0][2];
  }

  if (diagonalMarked(1, board)) {
    return board[0][0];
  }
  if (diagonalMarked(2, board)) {
    return board[2][0];
  }

  // if we get here, we need to check for any empty spot before we determine if game is a tie
  for (let row = 0; row < board.length; row++) {
    for (let col = 0; col < board[row].length; col++) {
      if (board[row][col] === null) {
        return null;
      }
    }
  }

  return TIE;
};

export const isUserVsUser = (settings: GameSettings) =>
  settings.playerX === PLAYER_TYPES.user &&
  settings.playerO === PLAYER_TYPES.user;

export const isComputerVsComputer = (settings: GameSettings) =>
  settings.playerX === PLAYER_TYPES.computer &&
  settings.playerO === PLAYER_TYPES.computer;

export const getUserValue = (settings: GameSettings) =>
  isUserVsUser(settings)
    ? PLAYER_X
    : settings.playerX === PLAYER_TYPES.user
    ? PLAYER_X
    : PLAYER_O;

export const getComputerValue = (settings: GameSettings) =>
  isComputerVsComputer(settings)
    ? PLAYER_X
    : settings.playerO === PLAYER_TYPES.computer
    ? PLAYER_O
    : PLAYER_X;
