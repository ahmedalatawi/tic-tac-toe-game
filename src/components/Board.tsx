import { View, StyleSheet } from "react-native";
import Square from "./Square";
import { GameBoard } from "../types/types";
import { GAME_STATES, PLAYER_X } from "../constants/common";

interface Props {
  board: GameBoard;
  gameState: string;
  disabled?: boolean;
  onPressUser: (rowIndex: number, colIndex: number) => void;
}

/**
 * This is a presentational (stateless) component for the game board
 *
 * @param param0
 * @returns
 */
function Board({ board, gameState, disabled, onPressUser }: Props) {
  return (
    <View style={styles.container}>
      {board.map((row, rowIndex) => (
        <View key={rowIndex}>
          {row.map((_, colIndex) => {
            const value = board[rowIndex][colIndex];

            return (
              <Square
                testID={`${rowIndex}${colIndex}`}
                key={colIndex}
                disabled={gameState === GAME_STATES.ended || disabled}
                onSquarePress={() => onPressUser(rowIndex, colIndex)}
                value={value ? (value === PLAYER_X ? "X" : "O") : null}
              />
            );
          })}
        </View>
      ))}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    paddingVertical: 10,
  },
});

export default Board;
