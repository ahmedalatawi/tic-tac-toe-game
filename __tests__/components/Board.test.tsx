import React from "react";
import renderer from "react-test-renderer";
import Board from "../../src/components/Board";
import board from "../__mocks__/boardData.mock";
import { render, screen, userEvent } from "@testing-library/react-native";
import { GAME_STATES, PLAYER_O, PLAYER_X } from "../../src/constants/common";

const onPressCallbackGlobal = jest.fn();

const mockProps = {
  board: board,
  gameState: GAME_STATES.inProgress,
  disabled: false,
  onPressUser: onPressCallbackGlobal,
};

describe("<Board />", () => {
  const BoardComponent = () => <Board {...mockProps} />;

  test("renders correctly", () => {
    const tree = renderer.create(<BoardComponent />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("return the correct row and col indices when first square pressed", async () => {
    render(<BoardComponent />);

    const firstSquare = screen.getByTestId("00");

    expect(firstSquare).toBeDefined();

    await userEvent.press(firstSquare);

    //enure the correct row and column indices are correct
    expect(onPressCallbackGlobal).toHaveBeenCalledWith(0, 0);
  });

  test("prevent pressing when board is disabled", async () => {
    const onPressCallback = jest.fn();
    render(
      <Board {...mockProps} disabled={true} onPressUser={onPressCallback} />,
    );

    const firstSquare = screen.getByTestId("00");

    expect(firstSquare).toBeDefined();

    await userEvent.press(firstSquare);

    //enure callBack did not fire
    expect(onPressCallback).not.toHaveBeenCalled();
  });

  test("should display X and O for PLAYER_X and PLAYER_O", async () => {
    const boardCopy = board.map((arr) => arr.slice());
    boardCopy[0][0] = PLAYER_X;
    boardCopy[0][1] = PLAYER_O;
    const onPressCallback = jest.fn();

    render(
      <Board {...mockProps} board={boardCopy} onPressUser={onPressCallback} />,
    );

    //Player X should exist on the board
    expect(screen.getByText("X")).toBeDefined();

    //Player O should exist on the board
    expect(screen.getByText("O")).toBeDefined();

    const squareX = screen.getByTestId("00");
    const squareO = screen.getByTestId("01");

    expect(squareX).toBeDefined();
    expect(squareO).toBeDefined();
  });

  test("should disable board when game is ended", async () => {
    const onPressCallback = jest.fn();
    render(
      <Board
        {...mockProps}
        gameState={GAME_STATES.ended}
        onPressUser={onPressCallback}
      />,
    );

    const square = screen.getByTestId("00");

    await userEvent.press(square);

    //enure board is disabled
    expect(onPressCallback).not.toHaveBeenCalled();
  });
});
