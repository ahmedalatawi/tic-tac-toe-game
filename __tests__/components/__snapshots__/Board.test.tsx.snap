// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> renders correctly 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "paddingVertical": 10,
    }
  }
>
  <View>
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 76,
          "justifyContent": "center",
          "width": 76,
        }
      }
      testID="00"
    >
      <Text
        style={
          {
            "fontSize": 40,
          }
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 76,
          "justifyContent": "center",
          "width": 76,
        }
      }
      testID="01"
    >
      <Text
        style={
          {
            "fontSize": 40,
          }
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 76,
          "justifyContent": "center",
          "width": 76,
        }
      }
      testID="02"
    >
      <Text
        style={
          {
            "fontSize": 40,
          }
        }
      />
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 76,
          "justifyContent": "center",
          "width": 76,
        }
      }
      testID="10"
    >
      <Text
        style={
          {
            "fontSize": 40,
          }
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 76,
          "justifyContent": "center",
          "width": 76,
        }
      }
      testID="11"
    >
      <Text
        style={
          {
            "fontSize": 40,
          }
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 76,
          "justifyContent": "center",
          "width": 76,
        }
      }
      testID="12"
    >
      <Text
        style={
          {
            "fontSize": 40,
          }
        }
      />
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 76,
          "justifyContent": "center",
          "width": 76,
        }
      }
      testID="20"
    >
      <Text
        style={
          {
            "fontSize": 40,
          }
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 76,
          "justifyContent": "center",
          "width": 76,
        }
      }
      testID="21"
    >
      <Text
        style={
          {
            "fontSize": 40,
          }
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 76,
          "justifyContent": "center",
          "width": 76,
        }
      }
      testID="22"
    >
      <Text
        style={
          {
            "fontSize": 40,
          }
        }
      />
    </View>
  </View>
</View>
`;
